# NumPy Exercises

## Exercise 1

- Import NumPy and recreate the following matrix
- Assign it to the variable `matrix_a`

array([[100, 103, 106, 109, 112, 115, 118, 121, 124, 127],
       [130, 133, 136, 139, 142, 145, 148, 151, 154, 157],
       [160, 163, 166, 169, 172, 175, 178, 181, 184, 187],
       [190, 193, 196, 199, 202, 205, 208, 211, 214, 217],
       [220, 223, 226, 229, 232, 235, 238, 241, 244, 247]])

## Exercise 2

Using a NumPy function, complete the following:
- Sum the rows and print the result
- Sum the columns and print the results

In [16]:
# these are the rows answers

array([1135, 1435, 1735, 2035, 2335])

In [18]:
# these are the columns answers

array([800, 815, 830, 845, 860, 875, 890, 905, 920, 935])

## Exercise 3

Create a matrix of 80 random integers from 100 to 200, and create a mask for the numbers above 150. Print the orginal matrix and the matrix with the mask.

In [19]:
# approximate matrix

array([[142, 127, 179, 164, 160, 138, 195, 124, 118, 171],
       [153, 184, 112, 184, 195, 110, 134, 194, 163, 155],
       [194, 124, 115, 123, 158, 177, 146, 184, 101, 180],
       [196, 152, 144, 124, 146, 128, 128, 108, 162, 107],
       [186, 193, 153, 133, 157, 129, 184, 162, 107, 197],
       [139, 199, 176, 168, 137, 140, 192, 173, 178, 176],
       [135, 107, 148, 173, 144, 135, 162, 157, 198, 123],
       [117, 197, 145, 146, 188, 165, 193, 193, 172, 155]])

In [22]:
# approximate array with mask applied

array([172, 159, 160, 197, 178, 193, 184, 183, 178, 163, 155, 153, 172,
       187, 182, 196, 153, 165, 153, 153, 173, 152, 171, 151, 168, 166,
       171, 198, 189, 183, 194, 180, 198, 187, 157, 157, 171])

## Exercise 4

Using the matrix you created in step three, create a mask with two conditions that show the numbers that are greater than 130 `and` the numbers that are lower than 170. Print the resulting array.

In [23]:
# possible array with mask

array([148, 138, 138, 145, 159, 134, 160, 145, 132, 163, 155, 153, 149,
       153, 143, 165, 145, 153, 153, 134, 152, 149, 133, 151, 144, 168,
       166, 131, 149, 148, 139, 134, 157, 157, 136, 148, 146])

## Exercise 5

Using the matrix you created in step three, create a mask with several conditions that show the numbers that are less than or equal to 125 `or` the numbers that are greater than or equal to 170. Print the resulting array.

In [24]:
# possible array with mask

array([172, 105, 107, 119, 119, 108, 197, 178, 124, 193, 184, 183, 178,
       122, 172, 187, 182, 124, 196, 118, 173, 123, 120, 171, 110, 118,
       171, 198, 106, 189, 183, 118, 194, 180, 198, 115, 187, 103, 105,
       171])

## Exercise 6

Using the matrix you created in step three, create a mask that shows only the **even** numbers. Print the resulting array.

In [25]:
# possible array with mask

array([148, 172, 138, 138, 126, 134, 160, 132, 108, 178, 124, 184, 126,
       178, 122, 172, 182, 124, 196, 118, 134, 152, 120, 110, 118, 144,
       168, 166, 198, 106, 118, 194, 180, 198, 148, 134, 136, 148, 146])

## Exercise 7

Using the matrix you created in step three, create a mask that shows only the **odd** numbers. Print the resulting array.

In [29]:
# possible array with mask

array([145, 159, 105, 107, 119, 119, 145, 197, 193, 183, 163, 155, 153,
       149, 187, 153, 143, 165, 145, 153, 153, 173, 123, 149, 127, 133,
       171, 151, 171, 131, 189, 183, 149, 139, 115, 187, 103, 105, 157,
       157, 171])

## Exercise 8

Using the matrix you created in step three, create a full copy of the matrix containing only every third column starting from the first one. Then create a mask that shows only the **odd** numbers. Print the resulting array.

In [33]:
# possible matrix with every third column

array([[148, 138, 159, 107],
       [160, 145, 197, 193],
       [184, 178, 153, 172],
       [187, 196, 165, 153],
       [153, 152, 120, 171],
       [110, 144, 171, 131],
       [189, 194, 198, 115],
       [187, 134, 136, 146]])

In [39]:
# possible array with the odd numbers of every third column

array([159, 107, 145, 197, 193, 153, 187, 165, 153, 153, 171, 171, 131,
       189, 115, 187])

## Exercise 9

- Unravel the matrix you created in step three.
- Reshape it into a differently-sized matrix
- Change te numbers at
    - third row, fourth column --> add 10
    - first row, last column --> to the power of 3
    - last row, second column --> subtract 20

## Exercise 10

Imagine you have 100 things you want to buy and that all of those cost the same amount of money. The total sum for the 100 things you want to buy is 100. Your friend gives you 50 and your sibling gives you 50 so that you can purchase all these things you want. You spend 97 and are left with 3. Then, you give 1 back to your friend and 1 back to your sibling and you stay with 1 yourself. If you now owe 49 to each one of them, where did the last dollar go?

- Create an array of 100 linearly spaced numbers that must add up to 100
- sum the array and assign it to a variable and call it `sum_100`
- create two variables with 50 in each and call them `debt1` and `debt2`
- subtract 1 from `debt1` and from `debt2`
- add `debt1` and `debt2` into `total_debt`
- subtract 97 from `sum_100` and assign it to a variable called `my_money`
- add up `my_money` and `total_debt`
- What happened?